[Feature] Add Differentiable JPEG

ChristophReich1996 opened the pull request:
#### Changes

As discussed with @shijianjian this pull request adds a differentiable JPEG encoding-decoding implementation to `kornia.enhance`. Differentiable JPEG coding can be used as a differentiable data augmentation approach or for optimizing JPEG for deep vision models. No new dependencies are introduced.

This implementation takes inspiration from the diff. JPEG approach of [Shin et al.](https://machine-learning-and-security.github.io/papers/mlsec17_paper_54.pdf) and [Reich et al.](https://arxiv.org/abs/2309.06978), however, uses a custom soft clipping operation (ELU-like). Additionally, the quantization tables are first clipped and then rounded (different from Reich et al.). For approximating the rounding function the polynomial approximation of Shin et al. is used. The implementation leads to similar results as the approach by Reich et al. (see [cardinality test](https://github.com/ChristophReich1996/kornia/blob/dc0f8248ba414a9dc9ed97dd0a216f45d587432f/test/enhance/test_jpeg.py#L86)).


#### Type of change
- [ ] 📚  Documentation Update
- [X] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [X] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] 📝 This change requires a documentation update (probably)

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+    KORNIA_CHECK_IS_TENSOR(quantization_table_y)
+    KORNIA_CHECK_IS_TENSOR(quantization_table_c)
+    # Check shape of inputs
+    KORNIA_CHECK_SHAPE(image_rgb, ["B", "3", "H", "W"])
```
would be great (or in a separated PR) to support out of the box non-batched images 3xHxW

ChristophReich1996 replied:
Let's do it in a separate PR if it's fine.

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+    dtype: Dtype = image_rgb.dtype
+    device: Device = image_rgb.device
+    # Quantization tables to same device and dtype as input image
+    quantization_table_y = quantization_table_y.to(dtype).to(device)
```
```suggestion
    quantization_table_y = quantization_table_y.to(dtype, device)
```

ChristophReich1996 replied:
Done :) https://github.com/kornia/kornia/blob/505765b226d4f7618c2812234a2193057f7df419/kornia/enhance/jpeg.py#L477-L478

edgarriba commented on test/enhance/test_jpeg.py:
```diff
@@ -0,0 +1,990 @@
+        assert img_jpeg is not None
+        assert img_jpeg.shape == img.shape
+
+    def test_gradcheck(self, device) -> None:
```
can you make somehow a test that performs a backward and check the gradients ? it would be funny to have a differentiable method without testing that part

ChristophReich1996 replied:
@edgarriba fully agree, however, PyTorch's grad check function is not really applicable here. I could check the gradient based on the reference implementation. What do you think about this?

ChristophReich1996 replied:
I added a test to check that the gradients are close to the reference implementation.

edgarriba commented on test/enhance/test_jpeg.py:
```diff
@@ -0,0 +1,990 @@
+            dtype=dtype,
+        )
+        # We use a slightly higher tolerance since our implementation varies from the reference implementation
+        assert_close(img_jpeg[..., :4, :4], img_jpeg_ref[..., :4, :4], rtol=0.01, atol=0.01)
```
```suggestion
        self.assert_close(img_jpeg[..., :4, :4], img_jpeg_ref[..., :4, :4], rtol=0.01, atol=0.01)
```
coming from `BaseTester`

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+    return y_encoded, cb_encoded, cr_encoded
+
+
+def _jpeg_decode(
```
would be interesting to include this one in the high level api as well (in another pr possibly)

ChristophReich1996 replied:
I don't see an immediate use for this function but if you like we can also include both `_jpeg_encode` and `_jpeg_decode`.

johnnv1 replied:
+1

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+        jpeg_quality,
+        quantization_table_y,
+    )
+    input_cb = _dequantize(
```
stupid question, why you cannot vectorise the process for the cb/cr channels in a single operation?

ChristophReich1996 replied:
Done in both the encoding and decoding.

ChristophReich1996 replied:
Very valid question btw ;)

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+    return rgb_decoded
+
+
+def diff_jpeg(
```
as the high level api for this module, i would like here to have a more explicit name like `encode_to_jpeg_differentiable`

ChristophReich1996 replied:
Agree, but `encode_to_jpeg_differentiable` is somewhat misleading since we perform both encoding and decoding. What about `jpeg_encode_decode_differentiable`?

ChristophReich1996 replied:
I change the naming to `jpeg_encode_decode_differentiable` and `JPEGEncodeDecodeDifferentiable` for the module. Let me know what you think :)

edgarriba replied:
got it -- in that case to have a more concise name `jpeg_codec_differentiable` where codec usually referred to compress/decompress

ChristophReich1996 replied:
Done :)

ChristophReich1996 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+        [49, 64, 78, 87, 103, 121, 120, 101],
+        [72, 92, 95, 98, 112, 100, 103, 99],
+    ],
+    dtype=torch.float,
```
Done :)

johnnv1 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,564 @@
+)
+from kornia.geometry import rescale
+
+QUANTIZATION_TABLE_Y: Tensor = torch.tensor(
```
move these tensors to be lazyloaded by putting them into a function. Otherwise, these tensors will be created when importing occurs; the function can have device and dtype as parameter, to initialize these tensor in the right place instead of moving/casting it after

ChristophReich1996 replied:
Not sure if correct but implemented like this:

https://github.com/kornia/kornia/blob/046443331956d8703683944e384cf21c75fb5d07/kornia/enhance/jpeg.py#L17-L32

johnnv1 replied:
yeah, perfect

ChristophReich1996 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+def diff_jpeg(
+    image_rgb: Tensor,
+    jpeg_quality: Tensor,
+    quantization_table_y: Tensor = QUANTIZATION_TABLE_Y,
```
Done

ChristophReich1996 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+    KORNIA_CHECK_IS_TENSOR,
+    KORNIA_CHECK_SHAPE,
+)
+from kornia.geometry import rescale
```
Done. 
Btw I was always wondering is there a reason why `__all__` is always a list and not a tuple? I know using a list is the default but wouldn't it make more sense to use an immutable object?

ChristophReich1996 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+        howerver, does not allow to excess the JPEG-coded byte file itself. For more details please refer to [1].
+
+    Reference:
+        [1] https://arxiv.org/abs/2309.06978
```
Done

ChristophReich1996 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,566 @@
+        required to first perform denormalization followed by normalizing the output images again.
+        Note, this implementation models the encoding-decoding mapping of JPEG in a differentiable setting,
+        howerver, does not allow to excess the JPEG-coded byte file itself. For more details please refer to [1].
+
```
Done

https://github.com/kornia/kornia/blob/7ca6f8e3a2c26d9972d2155af02e5ce0d4b2cade/kornia/enhance/jpeg.py#L525-L534

johnnv1 commented on test/enhance/test_jpeg.py:
```diff
@@ -0,0 +1,1001 @@
+        img_jpeg = diff_jpeg_module(img, jpeg_quality, qt_y, qt_c)
+        assert img_jpeg is not None
+        assert img_jpeg.shape == img.shape
+
+    def test_gradcheck(self, device) -> None:
```
```suggestion
    @pytest.mark.slow
    def test_gradcheck(self, device) -> None:
```

johnnv1 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+    KORNIA_CHECK_IS_TENSOR,
+    KORNIA_CHECK_SHAPE,
+)
+from kornia.geometry import rescale
```
```suggestion
from kornia.geometry.transform.affwarp import rescale
```

johnnv1 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+    )
+
+
+def _patchify_8x8(input: Tensor) -> Tensor:
```
would it be worth using https://kornia.readthedocs.io/en/latest/contrib.html#kornia.contrib.extract_tensor_patches and the `combine_tensor_patches` (instead of `_patchify_8x8` and ``_unpatchify_8x8) here? cc @edgarriba

ChristophReich1996 replied:
When I import `combine_tensor_patches` using `from kornia.contrib import extract_tensor_patches` I receive a circular import error 🤔

johnnv1 replied:
Try with the full path `from kornia.contrib.extract_patches import extract_tensor_patches`

ChristophReich1996 replied:
Unfortunately, loading `extract_patches` directly also leads to a circular import...

```
from kornia.enhance import normalize_min_max
E   ImportError: cannot import name 'normalize_min_max' from partially initialized module 'kornia.enhance' (most likely due to a circular import)
```

ChristophReich1996 replied:
Just importing kornia and calling `kornia.enhance.extract_tensor_patches` works...

johnnv1 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+    return output
+
+
+def _differentiable_clipping(
```
for a further pr, i think having `_differentiable_clipping`, ` _differentiable_polynomial_floor` and ` _differentiable_polynomial_rounding` into our utilities can be usefully :)

edgarriba replied:
open a ticket for that

johnnv1 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+    # Make DCT tensor and scaling
+    index: Tensor = torch.arange(8, dtype=dtype, device=device)
+    x, y, u, v = torch.meshgrid(index, index, index, index)
+    dct_tensor: Tensor = ((2.0 * x + 1.0) * u * torch.pi / 16.0).cos() * ((2.0 * y + 1.0) * v * torch.pi / 16.0).cos()
```
unrelated comment, if the torch has a `pi` constant, why do we have one too? cc @edgarriba

edgarriba replied:
historically wasn't there plus i remember some conflicts with the device api. Worth to revisit (open a ticket) and alias `[torch/kornia].pi`

ChristophReich1996 replied:
`torch.pi` directly links to `math.pi` I guess...

johnnv1 commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+       - :math:`QT_{c}` is the chroma quantization table.
+       - :math:`\hat{I}` is the resulting JPEG encoded-decoded image.
+
+    Notes:
```
```suggestion
    .. note::
```

ChristophReich1996 replied:
Done

johnnv1 reviewed (approved):
Thanks for this PR, it's really cool stuff. Just some minor comments

johnnv1 commented:
The CI is falling by an unrelated test, no worry, I'll merge #2745 which has a fix for it (I think so 😅 )

johnnv1 commented on test/enhance/test_jpeg.py:
```diff
@@ -0,0 +1,1002 @@
+import torch
+
+import kornia
+from kornia.testing import BaseTester
```
```suggestion
from testing.base import BaseTester
```

johnnv1 commented:
@edgarriba let me know when the changes are fine, I can squash it and put it on top of the main branch (CI is falling now because I updated to use the new testing, but just see now it is on top of the master instead of the main)

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+    return image_rgb_jpeg
+
+
+class JPEGEncodeDecodeDifferentiable(Module):
```
Not quite sure this is very valuable to provide since any pytorch user can easily create its own Module in any possible shape of the constructor. E.g if had to design the nn.Module of this function i'd put all the extra args (or maybe just a couple) into the module constructor so that the forward expect only the input image. But this is i guess of application specific, so i wouldn't support module (plus one less thing to maintain). Open to opinions @johnnv1 @shijianjian

ChristophReich1996 replied:
Good point! People might want to learn the quantization tables. So we could initialize the module with quantization tables. Let me know what you think about this. Maybe in a future PR?

ChristophReich1996 replied:
I added the option to specify the quantization tables (fixed and as parameters) in constructor, just in case someone wants to learn the quantization tables.

edgarriba commented on kornia/enhance/jpeg.py:
```diff
@@ -0,0 +1,573 @@
+    quantization_table_y: Tensor | None = None,
+    quantization_table_c: Tensor | None = None,
+) -> Tensor:
+    r"""Differentiable JPEG encoding-decoding module.
```
@ChristophReich1996 one more thing for future, is to include a meaningful visual output (auto-generated, or not) similar to other functions in this module

ChristophReich1996 replied:
Don't know what you mean by this... Do you mean adding an example image pair?

johnnv1 replied:
done

edgarriba reviewed (approved):
Left last comments but in general LGTM

ChristophReich1996 commented:
I added some more doc examples (use of QT's etc.), changed the module implementation (added an additional test for this), and addressed the comments. Let me know if there is still something to do. Thanks for all the feedback, always happy to learn something here :)

ChristophReich1996 commented:
Tests failed due to the changed import. Should I change it back?

![Screenshot 2024-01-24 at 21 07 51](https://github.com/kornia/kornia/assets/34400551/31f3605c-3bb0-47d1-8ede-d025dad98bc5)

johnnv1 commented:
> Tests failed due to the changed import. Should I change it back?

you can, but we will need to cherry-pick your changes and redo the PR to put it on top of the main, not the master. So will need to redo this import change

I can redo the PR later if you wish, but if you can do it that would be even better :)

(just changing the target branch will cause all commit history as conflict)

ChristophReich1996 commented:
Ah Ok as you like. I have run the test locally so I'm pretty sure they won't fail so we can also keep it this way...
